File: clients/shared/VertexCore/Sources/VertexCore/Transport/MQTTTransport.swift
import Foundation
import os

/// MQTT 5.0 transport for VPN tunnel data.
///
/// Runs on NWConnection (Network.framework) — safe for Network Extension use.
/// Handles broker failover, automatic reconnection, and resubscription.
/// Conforms to `Transport` (the abstract pub/sub contract used by
/// PacketTunnelProvider) so callers can swap the implementation later
/// without touching the data-plane code.
///
/// **Liveness model.** A long iteration through several iOS network APIs
/// (NEProvider.defaultPath KVO; process-wide NWPathMonitor; per-connection
/// viability/betterPath/pathUpdate handlers) showed each of them either
/// silent or noisy on at least one real-device scenario. The single
/// signal that consistently reflects reality on iPhone is application-
/// level keepalive: the broker is unreachable when MQTT PINGRESP doesn't
/// come back. Everything in this file is built around that one signal.
public final class MQTTTransport: Transport, @unchecked Sendable {
    private let queue = DispatchQueue(label: "ru.vertices.mqtt")
    private let logger = Logger(subsystem: "ru.vertices", category: "mqtt")

    // Configuration
    private var brokers: [BrokerURL]
    private let username: String
    private let password: String
    private let clientID: String

    // Tunables — see comments at use sites for the rationale.
    private static let connectTimeout: DispatchTimeInterval = .seconds(15)

    // State
    private var connection: MQTTConnection?
    // Pattern → handler map. Map (not list of pairs) so a second
    // subscribe(pattern:handler:) for the same pattern *replaces* the
    // handler instead of stacking it. See clients/PENDING_TRANSPORT_FIXES.md
    // "Bug #2": with a list of pairs, every reconnect-resubscribe path
    // (JoinManager retry, ControlChannel.reattach) appended yet another
    // copy of the same (pattern, handler) — N reconnects ⇒ N handler
    // calls per matching PUBLISH ⇒ exit log floods with N parallel
    // "Assigning IP" replies for one client.
    //
    // Insertion order is not preserved by Dictionary; if dispatch
    // ordering becomes load-bearing later, switch to OrderedDictionary
    // from swift-collections. Today no consumer relies on order —
    // JoinManager, ControlChannel, and ExitDiscovery each subscribe a
    // distinct pattern.
    private var subscriptions: [String: @Sendable (String, Data) -> Void] = [:]
    private var _isReady = false
    private var shouldReconnect = false
    private var currentBrokerIndex = 0
    private var reconnectAttempt = 0
    private var reconnectTimer: DispatchSourceTimer?
    private var connectTimeoutTimer: DispatchSourceTimer?
    /// Invoked when the link is declared dead (PINGRESP timeout, .failed
    /// on unsatisfied path, .waiting after ready) or after persistent
    /// connect-failure. The host's PacketTunnelProvider responds with
    /// `cancelTunnelWithError`, and the on-demand rule restarts the
    /// extension scoped to whatever interface iOS now considers best.
    /// Stored as a private var, only accessed under `queue` — the
    /// callback is supplied at init time so there's no cross-thread
    /// publish hazard on the var itself.
    private var onFatalError: (@Sendable (String) -> Void)?

    /// Invoked when the broker rejects our CONNECT (any non-zero CONNACK
    /// reason — bad creds, banned, server unavailable). Auth failures
    /// don't recover by retrying with the same credentials, so the
    /// transport stops reconnecting and lets the host surface a
    /// user-fixable error. Caller gets `(reasonCode, reasonString)`.
    private var onAuthFailure: (@Sendable (UInt8, String) -> Void)?

    /// Counts consecutive connect-attempt timeouts since the last CONNACK.
    /// After several in a row, escalate to `onFatalError` as a safety net
    /// when the immediate linkDead path didn't fire.
    private var consecutiveConnectFailures = 0

    // State stream for UI
    private var stateContinuation: AsyncStream<TransportState>.Continuation?
    private let _stateUpdates: AsyncStream<TransportState>

    // Continuation for waitReady
    private var readyContinuation: CheckedContinuation<Void, any Error>?

    /// True when connected and CONNACK received.
    public var isReady: Bool { queue.sync { _isReady } }

    /// Current broker host (or nil if disconnected).
    public var currentBroker: String? {
        queue.sync { _isReady ? brokers[currentBrokerIndex].host : nil }
    }

    /// Async stream of connection state changes.
    public var stateUpdates: AsyncStream<TransportState> { _stateUpdates }

    public init(
        brokers: [BrokerURL],
        username: String,
        password: String,
        clientID: String,
        onFatalError: (@Sendable (String) -> Void)? = nil,
        onAuthFailure: (@Sendable (UInt8, String) -> Void)? = nil
    ) {
        self.brokers = brokers
        self.username = username
        self.password = password
        self.clientID = clientID
        self.onFatalError = onFatalError
        self.onAuthFailure = onAuthFailure

        var cont: AsyncStream<TransportState>.Continuation!
        _stateUpdates = AsyncStream { cont = $0 }
        stateContinuation = cont
    }

    deinit {
        stateContinuation?.finish()
    }

    // MARK: - Lifecycle

    /// Connect to the first available broker. Returns when CONNACK received.
    public func start() async throws {
        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, any Error>) in
            queue.async { [weak self] in
                guard let self else {
                    cont.resume(throwing: CancellationError())
                    return
                }
                self.shouldReconnect = true
                self.readyContinuation = cont
                self.connectToCurrentBroker()
            }
        }
    }

    /// Graceful disconnect. Cancels reconnection. Resumes any pending
    /// `start()` continuation with `CancellationError` — leaking a
    /// `CheckedContinuation` is a Swift-runtime fatal error.
    public func stop() {
        queue.async { [weak self] in
            guard let self else { return }
            self.shouldReconnect = false
            self.cancelAllTimers()
            self.connection?.disconnect()
            self.connection = nil
            self._isReady = false
            self.readyContinuation?.resume(throwing: CancellationError())
            self.readyContinuation = nil
            self.emitState(.disconnected)
        }
    }

    /// Tear down the current connection and reconnect immediately. Right
    /// now this is only invoked internally from `handleConnectionEvent`
    /// when the underlying connection's `.disconnected` event fires
    /// (e.g. broker DISCONNECT, PINGRESP timeout). The Transport
    /// protocol exposes it for symmetry with future implementations.
    public func forceReconnect(reason: String) {
        queue.async { [weak self] in
            guard let self else { return }
            guard self.shouldReconnect else { return }
            self.logger.info("Force reconnect: \(reason, privacy: .public)")
            self._isReady = false
            self.connection?.disconnect()
            self.connection = nil
            self.reconnectAttempt = 0
            self.consecutiveConnectFailures = 0
            self.reconnectTimer?.cancel()
            self.reconnectTimer = nil
            self.cancelConnectTimeout()
            self.emitState(.reconnecting(broker: self.brokers[self.currentBrokerIndex].host, attempt: 0))
            self.connectToCurrentBroker()
        }
    }

    // MARK: - Test seams

    /// **Test-only.** Snapshot of the broker order in queue-safe form.
    /// Used by VertexCoreTests to assert that the Bug #1 demotion
    /// reordered the failover list after a CONNACK auth rejection.
    /// Production callers should observe broker selection through
    /// `currentBroker` / `state`.
    internal var _testBrokerHosts: [String] {
        queue.sync { brokers.map(\.host) }
    }

    /// **Test-only.** Synthesise a `.disconnected(connackReason:)` event
    /// on the transport's serial queue so a regression test for the
    /// sticky-vs-auth-failure interaction (PENDING_TRANSPORT_FIXES.md
    /// Bug #1) doesn't need a real broker. The synthesised event takes
    /// the same `handleConnectionEvent` path as a real CONNACK 0x86 from
    /// the broker.
    internal func _testFireAuthFailureDisconnect(connackReason: UInt8 = 0x86) {
        queue.sync {
            handleConnectionEvent(.disconnected(error: nil, linkDead: false, connackReason: connackReason))
        }
    }

    /// **Test-only.** Re-arm `onAuthFailure` between synthesised events.
    /// The production path nulls it out as part of the escalation
    /// shutdown, so the consecutive-demotion regression test needs a
    /// way to put it back without going through the public init.
    internal func _testReinstallAuthFailureCallback(_ cb: @escaping @Sendable (UInt8, String) -> Void) {
        queue.sync { onAuthFailure = cb }
    }

    /// **Test-only.** Synthesise an inbound PUBLISH and run the same
    /// dispatch path the wire reader would. Used by Bug #2 regression
    /// tests to assert that re-registering the same pattern replaces
    /// the handler instead of stacking.
    internal func _testInjectPublish(topic: String, payload: Data) {
        queue.sync {
            dispatchPublish(topic: topic, payload: payload)
        }
    }

    // MARK: - Publish

    /// Publish QoS 0. Drops silently when disconnected (fire-and-forget).
    /// retain=false is the only sensible value for VPN data plane traffic
    /// (a retained packet would replay on every new subscriber). Retained
    /// publishes belong on `Retainer` and are not used by current Swift
    /// callers — the exit (Go) is the only producer of retained messages.
    public func publish(topic: String, payload: Data) {
        queue.async { [weak self] in
            self?.connection?.publish(topic: topic, payload: payload, retain: false)
        }
    }

    // MARK: - Subscribe

    /// Register a subscription. Sends SUBSCRIBE immediately if connected
    /// **and** this is the first time this pattern is being registered;
    /// re-registering the same pattern just replaces the handler in
    /// place (no extra wire SUBSCRIBE — the broker already has the
    /// subscription, and the new handler will see the next PUBLISH).
    /// Patterns registered before connection are queued for the first
    /// connect's resubscribe.
    public func subscribe(
        pattern: String,
        handler: @escaping @Sendable (String, Data) -> Void
    ) {
        queue.async { [weak self] in
            guard let self else { return }
            let isNew = self.subscriptions[pattern] == nil
            self.subscriptions[pattern] = handler
            if self._isReady, isNew {
                self.connection?.subscribe(topics: [pattern])
            }
        }
    }

    /// Trigger an immediate liveness probe (PINGREQ) outside the regular
    /// keepalive cadence. Called when an external signal — system wake or
    /// a network path event — suggests the existing socket may be stale.
    /// On a dead link the probe's PINGRESP deadline (~5s) fires
    /// `linkDead`, escalating to extension restart far sooner than the
    /// next scheduled ping (~15s) would. No-op if not yet ready.
    public func checkLiveness(reason: String) {
        queue.async { [weak self] in
            guard let self else { return }
            guard self._isReady else {
                self.logger.info("checkLiveness(\(reason, privacy: .public)): not ready, skipping")
                return
            }
            self.connection?.pingNow(reason: reason)
        }
    }

    /// Wait until connected (with timeout).
    public func waitReady(timeout: Duration) async throws {
        if isReady { return }
        try await withThrowingTaskGroup(of: Void.self) { group in
            group.addTask {
                for await state in self.stateUpdates {
                    if case .connected = state { return }
                }
            }
            group.addTask {
                try await Task.sleep(for: timeout)
                throw TransportError.connectTimeout
            }
            try await group.next()
            group.cancelAll()
        }
    }

    // MARK: - Connection Management (private, runs on queue)

    private func connectToCurrentBroker() {
        guard shouldReconnect, !brokers.isEmpty else { return }

        let broker = brokers[currentBrokerIndex]
        logger.info("Connecting to \(broker.host, privacy: .public):\(broker.port, privacy: .public) (\(broker.scheme.rawValue, privacy: .public))")
        emitState(.connecting(broker: broker.host))

        let conn = MQTTConnection(
            clientID: clientID,
            username: username,
            password: password,
            queue: queue
        )

        conn.onStateChange = { [weak self] event in
            self?.handleConnectionEvent(event)
        }

        conn.onPublish = { [weak self] topic, payload in
            self?.dispatchPublish(topic: topic, payload: payload)
        }

        self.connection = conn
        scheduleConnectTimeout()
        conn.connect(to: broker)
    }

    /// Schedule a 15s deadline on reaching CONNACK. NWConnection by itself
    /// can sit in `.waiting` indefinitely (Apple Forums #768666 — known iOS
    /// 18 stick-in-waiting bug); without an application-side deadline we
    /// rely on the OS giving up, which it sometimes never does. On expiry
    /// we abort and let scheduleReconnect's backoff take over.
    private func scheduleConnectTimeout() {
        cancelConnectTimeout()
        let t = DispatchSource.makeTimerSource(queue: queue)
        t.schedule(deadline: .now() + Self.connectTimeout)
        t.setEventHandler { [weak self] in
            guard let self else { return }
            self.connectTimeoutTimer = nil
            guard !self._isReady, self.shouldReconnect else { return }
            self.consecutiveConnectFailures += 1
            self.logger.warning("Connect timeout — aborting (\(self.consecutiveConnectFailures, privacy: .public) consecutive failures)")
            self.connection?.disconnect()
            self.connection = nil
            // Fallback to extension restart only after persistent failure
            // — gives the path-monitor / requiredInterface mechanism time
            // to migrate naturally first.
            if self.consecutiveConnectFailures >= 3, let onFatal = self.onFatalError {
                self.logger.warning("Persistent connect failures — escalating to extension restart")
                self.shouldReconnect = false
                self.cancelAllTimers()
                self.onFatalError = nil
                onFatal("Persistent connect failures (\(self.consecutiveConnectFailures))")
                return
            }
            self.scheduleReconnect()
        }
        t.resume()
        connectTimeoutTimer = t
    }

    private func cancelConnectTimeout() {
        connectTimeoutTimer?.cancel()
        connectTimeoutTimer = nil
    }

    private func cancelAllTimers() {
        reconnectTimer?.cancel()
        reconnectTimer = nil
        cancelConnectTimeout()
    }

    private func handleConnectionEvent(_ event: MQTTConnection.ConnectionEvent) {
        // Already on queue (MQTTConnection callbacks run on our queue)
        switch event {
        case .connected:
            cancelConnectTimeout()
            logger.info("MQTT connected to \(self.brokers[self.currentBrokerIndex].host, privacy: .public)")
            _isReady = true
            reconnectAttempt = 0
            consecutiveConnectFailures = 0

            // Sticky reconnect: move successful broker to front
            if currentBrokerIndex > 0 {
                let winner = brokers.remove(at: currentBrokerIndex)
                brokers.insert(winner, at: 0)
                currentBrokerIndex = 0
            }

            // Resubscribe all patterns
            let topics = Array(subscriptions.keys)
            if !topics.isEmpty {
                connection?.subscribe(topics: topics)
                logger.info("Resubscribed \(topics.count, privacy: .public) topics")
            }

            emitState(.connected(broker: brokers[0].host))

            // Resume waitReady / start continuation
            readyContinuation?.resume()
            readyContinuation = nil

        case .disconnected(let error, let linkDead, let connackReason):
            cancelConnectTimeout()
            let wasReady = _isReady
            _isReady = false

            if let error {
                logger.error("MQTT disconnected: \(error.localizedDescription, privacy: .public) linkDead=\(linkDead, privacy: .public) connackReason=\(connackReason.map(String.init) ?? "nil", privacy: .public)")
            } else {
                logger.info("MQTT disconnected linkDead=\(linkDead, privacy: .public)")
            }

            // CONNACK rejection is a config issue (wrong creds, banned,
            // unsupported protocol). Retrying with the same credentials
            // will keep failing — demote the stickied broker, surface
            // to the host (if subscribed), and stop the loop. The
            // earlier `let onAuth = onAuthFailure` guard accidentally
            // gated the demotion on host subscription too — paritет с
            // Kotlin port, where demotion is pure routing state and
            // fires regardless of whether the host wants to be told.
            if let connackReason, connackReason != 0 {
                let reasonStr = (error as? MQTTCodecError).flatMap { e -> String? in
                    if case .connackFailed(let s) = e { return s } else { return nil }
                } ?? "code=\(connackReason)"
                logger.warning("CONNACK rejected (code=\(connackReason, privacy: .public)) — escalating, no retry")

                // Un-sticky: a previous successful connect promoted this broker
                // to index 0 (sticky reconnect). Auth-rejecting now means the
                // stickied broker is misconfigured (creds rotated, ACL
                // tightened, …). Demote it to the tail so a future start()
                // tries the original primary first — the user should not be
                // locked out by one bad broker once another is healthy.
                if brokers.count > 1, currentBrokerIndex == 0 {
                    let loser = brokers.remove(at: 0)
                    brokers.append(loser)
                    // Fresh start counters so the next start() begins its
                    // backoff cycle from 0 instead of resuming the old
                    // ladder against a different broker. Cosmetic — Windows
                    // doesn't reset either, but cleaner here.
                    reconnectAttempt = 0
                    consecutiveConnectFailures = 0
                }

                shouldReconnect = false
                cancelAllTimers()
                let cb = onAuthFailure
                onAuthFailure = nil
                onFatalError = nil
                cb?(connackReason, reasonStr)
                return
            }

            // Link-dead = physical interface we were bound to is gone
            // (PINGRESP timeout, .failed on unsatisfied path, .waiting
            // after ready). Same-process retries can't escape the
            // extension's interface scoping (verified empirically:
            // NWConnection rebind via NWParameters.requiredInterface
            // only takes the new path AFTER the path monitor has seen
            // the new default — which can be 5–10s). Escalate to a
            // full extension restart immediately so the on-demand rule
            // re-scopes to the new best path within ~1s.
            if linkDead, let onFatal = onFatalError {
                logger.warning("Link dead — escalating to extension restart")
                shouldReconnect = false
                cancelAllTimers()
                onFatalError = nil
                onFatal("Link dead")
                return
            }

            if wasReady {
                emitState(.reconnecting(broker: brokers[currentBrokerIndex].host, attempt: reconnectAttempt))
            }

            scheduleReconnect()
        }
    }

    private func scheduleReconnect() {
        guard shouldReconnect else { return }

        reconnectTimer?.cancel()

        // Advance to next broker
        reconnectAttempt += 1
        currentBrokerIndex = reconnectAttempt % brokers.count

        // Backoff: 0, 0.5, 1, 2, 5, 5, 5...
        let delays: [Double] = [0, 0.5, 1, 2, 5]
        let cycleIndex = min(reconnectAttempt / brokers.count, delays.count - 1)
        let delay = delays[cycleIndex]

        if delay == 0 {
            connectToCurrentBroker()
            return
        }

        logger.info("Reconnecting in \(delay, privacy: .public)s (attempt \(self.reconnectAttempt, privacy: .public), broker \(self.brokers[self.currentBrokerIndex].host, privacy: .public))")
        emitState(.reconnecting(broker: brokers[currentBrokerIndex].host, attempt: reconnectAttempt))

        let timer = DispatchSource.makeTimerSource(queue: queue)
        timer.schedule(deadline: .now() + delay)
        timer.setEventHandler { [weak self] in
            self?.connectToCurrentBroker()
        }
        timer.resume()
        reconnectTimer = timer
    }

    // MARK: - Dispatch incoming publishes

    private func dispatchPublish(topic: String, payload: Data) {
        for (pattern, handler) in subscriptions {
            if topicMatches(topic: topic, pattern: pattern) {
                handler(topic, payload)
            }
        }
    }

    /// MQTT topic matching: + = single level, # = multi-level (end only).
    private func topicMatches(topic: String, pattern: String) -> Bool {
        let topicParts = topic.split(separator: "/", omittingEmptySubsequences: false)
        let patternParts = pattern.split(separator: "/", omittingEmptySubsequences: false)

        var ti = 0, pi = 0
        while pi < patternParts.count {
            let pp = patternParts[pi]
            if pp == "#" { return true } // matches rest
            guard ti < topicParts.count else { return false }
            if pp != "+" && pp != topicParts[ti] { return false }
            ti += 1
            pi += 1
        }
        return ti == topicParts.count
    }

    // MARK: - State emission

    private func emitState(_ state: TransportState) {
        stateContinuation?.yield(state)
    }
}

File: clients/android/Vertex/core/src/main/kotlin/ru/vertices/android/core/mqtt/MqttTransport.kt
package ru.vertices.android.core.mqtt

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import ru.vertices.android.core.config.BrokerUrl
import ru.vertices.android.core.protocol.Topics
import timber.log.Timber
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

/** Coarse transport state for the UI. Mirror of Swift `TransportState`. */
sealed interface TransportState {
    data object Disconnected : TransportState
    data class Connecting(val broker: String) : TransportState
    data class Connected(val broker: String) : TransportState
    data class Reconnecting(val broker: String, val attempt: Int) : TransportState
}

/**
 * MQTT 5.0 transport. Mirror of Swift `MQTTTransport` — multi-broker failover
 * with sticky reconnect, a small fixed backoff, and resubscribe on each CONNACK.
 *
 * Threading: all state mutation happens on the scheduler's single thread. Public
 * methods marshal onto it. The state flow can be observed from any thread.
 *
 * Liveness model: a single application-level signal — PINGRESP timeout — drives
 * "link dead" detection. Same reasoning as iOS: every other system network API
 * (NWPathMonitor, default-path KVO) was unreliable on at least one real device
 * scenario. Android `ConnectivityManager.NetworkCallback` is consulted for hints
 * (Phase 2+), but PINGRESP is authoritative.
 */
class MqttTransport(
    initialBrokers: List<BrokerUrl>,
    private val username: String,
    private val password: String,
    private val clientId: String,
    private val keepAliveSeconds: Int = 20,
    private val socketProtector: SocketProtector? = null,
    private val socketFactory: (BrokerUrl) -> MqttSocket = { defaultSocketFor(it, socketProtector) },
    private val scheduler: ScheduledExecutorService = defaultScheduler(),
    private val onFatalError: ((reason: String) -> Unit)? = null,
    private val onAuthFailure: ((reasonCode: Int, reasonString: String) -> Unit)? = null,
) {

    private val brokers: MutableList<BrokerUrl> = ArrayList(initialBrokers)
    init { require(brokers.isNotEmpty()) { "at least one broker required" } }

    // pattern → handler. Map (not list of pairs) so a second `subscribe` for the
    // same pattern replaces the handler instead of stacking — otherwise dispatch
    // would fan a single broker payload out to every accumulated handler. The
    // race that motivated the change: subscribe arrived between `ready=true`
    // and the resubscribe loop in `handleConnectionEvent.Connected`, ending up
    // double-registered in the previous list-based representation.
    // LinkedHashMap to preserve insertion order in `dispatch` for determinism.
    private val subscriptions: LinkedHashMap<String, (String, ByteArray) -> Unit> = LinkedHashMap()

    private val _state = MutableStateFlow<TransportState>(TransportState.Disconnected)
    val state: StateFlow<TransportState> = _state.asStateFlow()

    @Volatile private var ready = false
    @Volatile private var shouldReconnect = false
    @Volatile private var connection: MqttConnection? = null
    private var currentBrokerIndex = 0
    private var reconnectAttempt = 0
    private var consecutiveConnectFailures = 0
    private var connectTimeoutTask: ScheduledFuture<*>? = null
    private var reconnectTask: ScheduledFuture<*>? = null

    val isReady: Boolean get() = ready
    val currentBroker: BrokerUrl? get() = if (ready) brokers.getOrNull(currentBrokerIndex) else null

    fun start() {
        scheduler.execute {
            shouldReconnect = true
            connectToCurrentBroker()
        }
    }

    fun stop() {
        scheduler.execute {
            shouldReconnect = false
            cancelAllTimers()
            connection?.disconnect()
            connection = null
            ready = false
            _state.value = TransportState.Disconnected
        }
    }

    fun publish(topic: String, payload: ByteArray) {
        // Fire-and-forget — drop if not ready (caller retries via app-level keepalive).
        scheduler.execute {
            connection?.publish(topic, payload, retain = false, messageExpirySeconds = 10)
        }
    }

    fun subscribe(pattern: String, handler: (String, ByteArray) -> Unit) {
        scheduler.execute {
            val isNew = !subscriptions.containsKey(pattern)
            subscriptions[pattern] = handler
            // Only ask the broker to subscribe the first time we see the pattern;
            // a re-subscribe with the same filter is harmless on MQTT 5 but
            // doubles message dispatch when paired with a list-based local map.
            if (ready && isNew) connection?.subscribe(listOf(pattern))
        }
    }

    /**
     * Stop dispatching messages for [pattern] locally. We do not send an MQTT
     * UNSUBSCRIBE today (the codec doesn't encode it yet — Phase 2); the broker
     * will keep delivering, but [dispatch] becomes a no-op for this filter, so
     * the cost is just one ignored frame per heartbeat / control message.
     */
    fun unsubscribe(pattern: String) {
        scheduler.execute {
            subscriptions.remove(pattern)
        }
    }

    /** Trigger a fresh PINGREQ ahead of cadence. Used on network change hints. */
    fun checkLiveness() {
        scheduler.execute { connection?.pingNow() }
    }

    // ===== Test seams (internal — accessible only inside the :core module) =====

    /**
     * **Test-only.** Snapshot the broker order from the scheduler thread.
     * Used by [MqttTransportBug1Test] to assert that the Bug #1 demotion
     * reordered the failover list after a CONNACK auth rejection.
     * Production callers should observe broker selection through
     * [currentBroker] / [state].
     */
    internal fun _testBrokerHosts(): List<String> =
        scheduler.submit<List<String>> { brokers.map { it.host } }.get()

    /**
     * **Test-only.** Synthesise a `MqttConnection.Event.Disconnected`
     * carrying [connackReason] on the transport's scheduler thread —
     * the same thread the real listener callback uses. Lets a regression
     * test for the sticky-vs-auth-failure interaction
     * (clients/PENDING_TRANSPORT_FIXES.md "Bug #1") run without standing
     * up a real broker.
     */
    internal fun _testFireAuthFailureDisconnect(connackReason: Int = 0x86) {
        scheduler.submit {
            handleConnectionEvent(MqttConnection.Event.Disconnected(
                cause = null, linkDead = false, connackReason = connackReason))
        }.get()
    }

    /**
     * **Test-only.** Force [currentBrokerIndex] for tests that want to
     * exercise behaviour while connected to a non-primary (e.g. "auth
     * failure on the backup must NOT demote the primary it never
     * touched"). Real production transitions go through `Connected`'s
     * sticky-promote path.
     */
    internal fun _testSetCurrentBrokerIndex(index: Int) {
        scheduler.submit { currentBrokerIndex = index }.get()
    }

    fun forceReconnect(reason: String) {
        scheduler.execute {
            if (!shouldReconnect) return@execute
            Timber.tag(TAG).i("force reconnect: $reason")
            ready = false
            connection?.disconnect()
            connection = null
            reconnectAttempt = 0
            consecutiveConnectFailures = 0
            cancelAllTimers()
            _state.value = TransportState.Reconnecting(brokers[currentBrokerIndex].host, 0)
            connectToCurrentBroker()
        }
    }

    // ---- internal ----

    private fun connectToCurrentBroker() {
        if (!shouldReconnect) return
        val b = brokers[currentBrokerIndex]
        Timber.tag(TAG).i("connecting to ${b.host}:${b.port} (${b.scheme.raw})")
        _state.value = TransportState.Connecting(b.host)

        val sock = socketFactory(b)
        val conn = MqttConnection(
            socket = sock,
            clientId = clientId,
            username = username,
            password = password,
            keepAliveSeconds = keepAliveSeconds,
            scheduler = scheduler,
        )
        connection = conn
        scheduleConnectTimeout()

        conn.connect(
            listener = { handleConnectionEvent(it) },
            onPublish = { topic, payload -> dispatch(topic, payload) },
        )
    }

    private fun handleConnectionEvent(event: MqttConnection.Event) {
        when (event) {
            MqttConnection.Event.Connected -> {
                cancelConnectTimeout()
                Timber.tag(TAG).i("connected to ${brokers[currentBrokerIndex].host}")
                ready = true
                reconnectAttempt = 0
                consecutiveConnectFailures = 0

                // Sticky reconnect: move winner to front so next reconnect tries it first.
                if (currentBrokerIndex > 0) {
                    val winner = brokers.removeAt(currentBrokerIndex)
                    brokers.add(0, winner)
                    currentBrokerIndex = 0
                }
                // Resubscribe everything we have on file.
                val patterns = subscriptions.keys.toList()
                if (patterns.isNotEmpty()) {
                    connection?.subscribe(patterns)
                    Timber.tag(TAG).i("resubscribed ${patterns.size} topics")
                }
                _state.value = TransportState.Connected(brokers[0].host)
            }
            is MqttConnection.Event.Disconnected -> {
                cancelConnectTimeout()
                val wasReady = ready
                ready = false
                connection = null
                val cause = event.cause
                Timber.tag(TAG).w(cause, "MQTT disconnected linkDead=${event.linkDead} connackReason=${event.connackReason}")

                // Auth/CONNACK rejection: short-circuit. Same creds will keep failing.
                val rc = event.connackReason
                if (rc != null && rc != 0) {
                    Timber.tag(TAG).w("CONNACK rejected (code=$rc) — escalating, no retry")

                    // Un-sticky: a previous successful connect promoted this
                    // broker to index 0 (sticky reconnect). Auth-rejecting now
                    // means the stickied broker is misconfigured (creds rotated,
                    // ACL tightened, …). Demote it to the tail so a future
                    // start() tries the original primary first — the user
                    // shouldn't be locked out by one bad broker once another
                    // is healthy. See clients/PENDING_TRANSPORT_FIXES.md "Bug #1".
                    if (brokers.size > 1 && currentBrokerIndex == 0) {
                        val loser = brokers.removeAt(0)
                        brokers.add(loser)
                        // Fresh start counters so the next start() begins its
                        // backoff cycle from 0 (paritет с iOS / macOS fix).
                        reconnectAttempt = 0
                        consecutiveConnectFailures = 0
                    }

                    shouldReconnect = false
                    cancelAllTimers()
                    onAuthFailure?.invoke(rc, ReasonStrings.connack(rc))
                    return
                }

                // linkDead means the underlying network path died: PINGRESP
                // timeout fired, or socket I/O threw post-handshake. The TUN
                // keeps running — only the MQTT transport needs to rebuild
                // on whatever path is now default. Mirrors iOS, where the
                // extension cancels itself with `cancelTunnelWithError` and
                // the on-demand rule restarts it; Android has no on-demand
                // equivalent so we reconnect in-place. Previously this path
                // escalated to fatal, killing the whole tunnel on every
                // wifi/cell handoff and stranding the user with no auto-
                // reconnect.
                if (event.linkDead) {
                    Timber.tag(TAG).i("link dead — reconnecting")
                }

                if (wasReady) {
                    _state.value = TransportState.Reconnecting(brokers[currentBrokerIndex].host, reconnectAttempt)
                }
                scheduleReconnect()
            }
        }
    }

    private fun scheduleConnectTimeout() {
        cancelConnectTimeout()
        connectTimeoutTask = scheduler.schedule({
            if (!ready && shouldReconnect) {
                consecutiveConnectFailures++
                Timber.tag(TAG).w("connect timeout — aborting ($consecutiveConnectFailures consecutive)")
                connection?.disconnect()
                connection = null
                if (consecutiveConnectFailures >= 3) onFatalError?.let {
                    shouldReconnect = false
                    cancelAllTimers()
                    it("Persistent connect failures ($consecutiveConnectFailures)")
                    return@schedule
                }
                scheduleReconnect()
            }
        }, CONNECT_TIMEOUT_SEC, TimeUnit.SECONDS)
    }

    private fun cancelConnectTimeout() {
        connectTimeoutTask?.cancel(false); connectTimeoutTask = null
    }

    private fun cancelAllTimers() {
        cancelConnectTimeout()
        reconnectTask?.cancel(false); reconnectTask = null
    }

    private fun scheduleReconnect() {
        if (!shouldReconnect) return
        reconnectTask?.cancel(false)

        reconnectAttempt++
        currentBrokerIndex = reconnectAttempt % brokers.size

        val cycleIndex = (reconnectAttempt / brokers.size).coerceAtMost(BACKOFF_DELAYS_SEC.size - 1)
        val delaySec = BACKOFF_DELAYS_SEC[cycleIndex]

        if (delaySec == 0.0) {
            connectToCurrentBroker(); return
        }

        Timber.tag(TAG).i("reconnect in ${delaySec}s (attempt=$reconnectAttempt, broker=${brokers[currentBrokerIndex].host})")
        _state.value = TransportState.Reconnecting(brokers[currentBrokerIndex].host, reconnectAttempt)
        reconnectTask = scheduler.schedule(::connectToCurrentBroker,
            (delaySec * 1000).toLong(), TimeUnit.MILLISECONDS)
    }

    private fun dispatch(topic: String, payload: ByteArray) {
        // Snapshot to avoid ConcurrentModificationException — `subscribe` /
        // `unsubscribe` mutate `subscriptions` from the same scheduler thread,
        // but dispatch is invoked from MqttConnection's read callback which
        // can fire while a subscribe is in-flight in a queued runnable.
        val snapshot = subscriptions.entries.toList()
        for ((pattern, handler) in snapshot) {
            if (Topics.matches(topic, pattern)) {
                handler(topic, payload)
            }
        }
    }

    companion object {
        private const val TAG = "vtx-mqtt-tx"
        private const val CONNECT_TIMEOUT_SEC = 15L
        private val BACKOFF_DELAYS_SEC = doubleArrayOf(0.0, 0.5, 1.0, 2.0, 5.0)

        private fun defaultScheduler(): ScheduledExecutorService =
            Executors.newSingleThreadScheduledExecutor { r ->
                Thread(r, "vtx-mqtt-tx").apply { isDaemon = true }
            }

        private fun defaultSocketFor(b: BrokerUrl, protector: SocketProtector?): MqttSocket =
            if (b.isWebSocket) WssMqttSocket(b) else TlsMqttSocket(b, protector)
    }
}
